lib/collection/src/hash_ring.rs

```diff
@@ -195,105 +176,188 @@ impl<T: Hash + Copy + PartialEq> HashRingRouter<T> {
+            }
+
+            HashRing::Fair { ring, scale, .. } => {
+                for idx in 0..*scale {
+                    ring.add((shard, idx));
                 }
             }
         }
+
+        true
     }
 
     pub fn remove(&mut self, shard: &T) -> bool {
+        if !self.nodes_mut().remove(shard) {
+            return false;
+        }
+
         match self {
-            HashRing::Raw(ring) => ring.remove(shard).is_some(),
-            HashRing::Fair { ring, scale } => {
-                let mut removed = false;
-                for i in 0..*scale {
-                    if ring.remove(&(*shard, i)).is_some() {
-                        removed = true;
-                    }
+            HashRing::Raw { ring, .. } => {
+                ring.remove(shard);
+            }
+
+            HashRing::Fair { ring, scale, .. } => {
+                for idx in 0..*scale {
+                    ring.remove(&(*shard, idx));
                 }
-                removed
             }
         }
+
+        true
     }
+}
 
+impl<T: Eq + Hash> HashRing<T> {
     pub fn get<U: Hash>(&self, key: &U) -> Option<&T> {
         match self {
-            HashRing::Raw(ring) => ring.get(key),
+            HashRing::Raw { ring, .. } => ring.get(key),
             HashRing::Fair { ring, .. } => ring.get(key).map(|(shard, _)| shard),
         }
     }
 
     pub fn is_empty(&self) -> bool {
+        self.nodes().is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.nodes().len()
+    }
+
+    pub fn contains(&self, shard: &T) -> bool {
+        self.nodes().contains(shard)
+    }
+
+    pub fn nodes(&self) -> &HashSet<T> {
         match self {
-            HashRing::Raw(ring) => ring.is_empty(),
-            HashRing::Fair { ring, .. } => ring.is_empty(),
+            HashRing::Raw { nodes, .. } => nodes,
+            HashRing::Fair { nodes, .. } => nodes,
         }
     }
 
-    pub fn len(&self) -> usize {
+    fn nodes_mut(&mut self) -> &mut HashSet<T> {
         match self {
-            HashRing::Raw(ring) => ring.len(),
-            HashRing::Fair { ring, scale } => ring.len() / *scale as usize,
+            HashRing::Raw { nodes, .. } => nodes,
+            HashRing::Fair { nodes, .. } => nodes,
         }
     }
-}
 
-impl<T: Hash + Copy + PartialEq + Eq> HashRing<T> {
-    /// Get unique nodes from the hashring
-    pub fn unique_nodes(&self) -> HashSet<T> {
+    fn scale(&self) -> Option<u32> {
         match self {
-            HashRing::Raw(ring) => ring.clone().into_iter().collect(),
-            HashRing::Fair { ring, .. } => ring.clone().into_iter().map(|(node, _)| node).collect(),
+            HashRing::Raw { .. } => None,
+            HashRing::Fair { scale, .. } => Some(*scale),
         }
     }
 }
 
+impl<T: Eq + Hash> PartialEq for HashRing<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.scale() == other.scale() && self.nodes() == other.nodes()
+    }
+}
+
+impl<T: Eq + Hash> Eq for HashRing<T> {}
```

Comment:
Can't we just derive PartialEq here as well? I believe that should be fine now since this made add/remove idempotent by the attached nodes hash set.

Reply:
Hmmmm.... You're right, it should be safe now.

Reply:
Done.